### PR TITLE
[Core] Remove pure-python `Specification` implementation

### LINKED ISSUE #342: Remove old `Specifications` definitions/code
- Remove `specifications.py`, `SpecificationFactory.py` and `Specification.py` along with corresponding tests.
- Remove doc references to `(Entity|Locale|Relationship)Specification`.
- Migrate test harness to manually implemented Python Specification/Traits

### PR DESCRIPTION
**Pull request by foundrytom**
Removes the old pure-python `Specification` implementation, and moves code over to using the new C++ `Specification`/`Traits` system.

Closes #342.

NB. There may look to be a lot of commits, but they are all fairly small and isolated to a specific type of change.

### CODE REVIEW

**Review by feltech — APPROVED** [review]
I'm still checking the codebase to see if I can find anything that's been missed, but so far so good.

File: python/openassetio/Trait.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+#
```
[the remark this answers is not in the document]
Comment (foundrytom) [reply python/openassetio/Trait.py:1]: It is indeed that thing!

File: python/openassetio/test/manager/specifications.py [hunk 2]
```diff
@@ -18,11 +18,29 @@
...
 
+from openassetio import Trait
+
 from ..specifications import TestHarnessLocale
 
 
+# pylint: disable=missing-function-docstring
+
+
+class ManagerTrait(Trait):
```
[the remark this answers is not in the document]
Comment (foundrytom) [reply python/openassetio/test/manager/specifications.py:29]: Yeah, its a good point. Happy to `[plan]` to sort this out in another PR?

File: python/openassetio/Context.py [hunk 3]
```diff
@@ -233,11 +235,11 @@ def __setLocale(self, locale):
     ## than a 'Write node' from a comp tree could result in different
     ## behavior.
     ##
-    ## The Locale uses a @ref openassetio.specifications.LocaleSpecification
-    ## to describe in more detail, what specific part of a @ref host
-    ## is requesting an action. In the case of a file browser for example,
-    ## it may also include information such as whether or not multi-selection
-    ## is required.
+    ## The Locale uses a @fqref{specification.Specification}
```
[the remark this answers is not in the document]
Comment (foundrytom) [reply python/openassetio/Context.py]: Updated [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/350/commits/b2c919c7337ea1e41873b8a8a1e3415281da10f1).

File: python/openassetio/test/manager/specifications.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (feltech) [reply python/openassetio/test/manager/specifications.py:29]: Yeah that's fine

**Comment by feltech** [comment]
I've scanned through the codebase and can't find anything that isn't migrated, at least considering #249 too.  Happy for this to be squashed and merged.

File: python/openassetio/test/manager/specifications.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (foundrytom) [reply python/openassetio/test/manager/specifications.py:29]: https://github.com/TheFoundryVisionmongers/OpenAssetIO/issues/355

**Pull request merged**